### PR TITLE
samples(storage): add samples for object contexts

### PR DESCRIPTION
**Pull request by krishnamd-jkp**
## Description

Fixes #<ISSUE-NUMBER>

Note: Before submitting a pull request, please open an issue for discussion if you are not associated with Google.

## Checklist
- [x] I have followed [Contributing Guidelines from CONTRIBUTING.MD](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/CONTRIBUTING.md)
- [X] **Tests** pass:   `go test -v ./..` (see [Testing](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/CONTRIBUTING.md#testing))
- [X] **Code formatted**:   `gofmt` (see [Formatting](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/CONTRIBUTING.md#formatting))
- [X] **Vetting** pass:   `go vet` (see [Formatting](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/CONTRIBUTING.md#formatting))
- [ ] These samples need a new **API enabled** in testing projects to pass (let us know which ones)
- [ ] These samples need a new/updated **env vars** in testing projects set to pass (let us know which ones)
- [ ] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/.github/CODEOWNERS) with the codeowners for this sample
- [ ] This sample adds a new **Product API**, and I updated the [Blunderbuss issue/PR auto-assigner](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/.github/blunderbuss.yml) with the codeowners for this sample
- [X] Please **merge** this PR for me once it is approved

### CODE REVIEW

File: storage/objects/set_object_contexts.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+	ctx, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+
+	o := client.Bucket(bucket).Object(object)
+
+	// To set contexts on a new object during write (followed by writing
+	// data and closing the writer to set contexts):
+	//
+	// writer := o.NewWriter(ctx)
```
Comment (tritone) [inline storage/objects/set_object_contexts.go]: Generally having this much commented out code (here and below) in samples is not ideal, since it's not able to be checked that it will compile and work.

Can we do some of this inline, and/or remove it from the sample? What do other languages do?

Comment (krishnamd-jkp) [reply storage/objects/set_object_contexts.go]: This is what Java SDK did. They covered different ways to patch contexts in the same sample with commented code. I feel it is okay as all the examples are cohesive. Wdyt?

File: go.mod
Comment (cpriti-os) [inline go.mod:1]: Please verify if we need any dependency changes outside of the directory which is changed, in this case storage/objects

File: storage/objects/get_object_contexts.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	ctx, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+
+	o := client.Bucket(bucket).Object(object)
+	attrs, err := o.Attrs(ctx)
+	if err != nil {
+		return fmt.Errorf("Object(%q).Attrs: %w", object, err)
+	}
+
+	fmt.Fprintf(w, "Object contexts for %v:\n", attrs.Name)
```
Comment (cpriti-os) [inline storage/objects/get_object_contexts.go]: This statement should be inside the below if?

File: storage/objects/list_object_contexts.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	 * filter := "contexts.\"KEY\":*";
+	 *
+	 * List any object that does not have a context with the specified key attached
+	 * filter := "-contexts.\"KEY\":*";
+	 *
+	 * List any object that has a context with the specified key and value attached
+	 * filter := "contexts.\"KEY\"=\"VALUE\"";
+	 *
+	 * List any object that does not have a context with the specified key and value attached
+	 * filter := "-contexts.\"KEY\"=\"VALUE\"";
```
Comment (cpriti-os) [inline storage/objects/list_object_contexts.go:45]: Could you check if we can link a public documentation to explain the different ways we can provide these filters?

File: storage/objects/objects_test.go [hunk 4]
```diff
@@ -842,3 +842,50 @@ func TestListSoftDeletedVersionsOfObject(t *testing.T) {
...
+	}
+	t.Cleanup(func() { client.Close() })
+
+	bucketName := testutil.CreateTestBucket(ctx, t, client, tc.ProjectID, testPrefix)
+	objectName := "context-object.txt"
+
+	if err := uploadFile(io.Discard, bucketName, objectName); err != nil {
+		t.Fatalf("uploadFile(%q): %v", objectName, err)
+	}
+
```
Comment (cpriti-os) [inline storage/objects/objects_test.go]: Can you add a test where get object contexts returns no context (Since there are none)

File: storage/objects/set_object_contexts.go
Comment (cpriti-os) [inline storage/objects/set_object_contexts.go:1]: Shouldn't we also add a sample for deleting contexts?

File: storage/objects/set_object_contexts.go (hunk 1 above)
Comment (cpriti-os) [reply storage/objects/set_object_contexts.go]: I agree with Chris here, providing sample code in comments does not serve the purpose, why not set two contexts, first while writing the object and second after object is created.

File: storage/objects/set_object_contexts.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	objectAttrsToUpdate := storage.ObjectAttrsToUpdate{
+		Contexts: &storage.ObjectContexts{
+			Custom: map[string]storage.ObjectCustomContextPayload{
+				"key1": {Value: "value1"},
+				"key2": {Value: "value2"},
+			},
+		},
+	}
+
+	// To delete all existing contexts:
```
Comment (cpriti-os) [inline storage/objects/set_object_contexts.go]: This should be a separate sample IMO

File: storage/objects/objects_test.go (hunk 4 above)
Comment (krishnamd-jkp) [reply storage/objects/objects_test.go]: Added the test. The modified test tests all the scenarios. Kind of a long test but it shows the lifecycle of object contexts together making it cohesive.

File: storage/objects/delete_object_contexts.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package objects
+
+// [START storage_delete_object_contexts]
```
Comment (cpriti-os) [inline storage/objects/delete_object_contexts.go:17]: Ensure we are using the existing region tags for these samples if available, rest looks fine.

Comment (krishnamd-jkp) [reply storage/objects/delete_object_contexts.go:17]: Done. Reused list region tags. Others are new.

File: storage/objects/patch_object_contexts.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package objects
+
+// [START storage_patch_object_contexts]
```
Comment (cpriti-os) [inline storage/objects/patch_object_contexts.go]: use set object context tag here, feel free to rename the file as well if required.
